p4-16/spec/P4-16-spec.mdk

```diff
@@ -7716,33 +7716,54 @@ The evaluation of a P4 program is done in two stages:
-- The following expressions (`+`, `-`, `*`, `/ `, `%`, `!`, `&`, `|`, `&&`, `||`, `<< `, `>> `, `~` `, `\ `>`, `<`, `==`, `!=`, `<=`, `>=`, `++`, `[:]`, `?:`)
-  when their operands are all compile-time known values.
+- The expression `{#}` representing an invalid header or header union value.
+- Instances constructed by instance declarations (Section [#sec-instantiations]) and constructor invocations.
+- Identifiers that represent declared types, actions, tables, parsers, controls, or packages.
+- Tuple expression where all components are local compile-time known values.
+- Structure-valued expressions, where all fields are local compile-time known values.
+- Expressions evaluating to a list type, where all elements are local compile-time known values.
+- Legal casts applied to local compile-time known values.
+- The following expressions (`+`, `-`, `|+|`, `|-|`, `*`, `/ `, `%`, `!`, `&`, `|`, `^`, `&&`, `||`, `<< `, `>>`, `~`, `/`, `>`, `<`, `==`, `!=`, `<=`, `>=`, `++`, `[:]`, `?:`) when their operands are all local compile-time known values.
+- Expressions of the form `e.minSizeInBits()`, `e.minSizeInBytes()`, `e.maxSizeInBits()` and `e.maxSizeInBytes()` where the type of `e` is not generic.
+
+The following are compile-time known values:
+
+- All local compile-time known values.
+- Constructor parameters.
 - Identifiers declared as constants using the `const` keyword.
-- Expressions of the form `e.minSizeInBits()`, `e.minSizeInBytes()`,
-  `e.maxSizeInBits()` and `e.maxSizeInBytes()`
+- Tuple expression where all components are compile-time known values.
+- Expressions evaluating to a list type, where all elements are compile-time known values.
+- Structure-valued expressions, where all fields are compile-time known values.
+- Expressions evaluating to a list type, where all elements are compile-time known values.
+- Legal casts applied to compile-time known values.
+- The following expressions (`+`, `-`, `|+|`, `|-|`, `*`, `/ `, `%`, `cast`, `!`, `&`, `|`, `^`, `&&`, `||`, `<< `, `>> `, `~`, `/`, `>`, `<`, `==`, `!=`, `<=`, `>=`, `++`, `[:]`, `?:`) when their operands are all compile-time known values.
+- Expressions of the form `e.minSizeInBits()`, `e.minSizeInBytes()`, `e.maxSizeInBits()` and `e.maxSizeInBytes()` where the the type of `e` is generic.
+
+Intuitively, the main difference between _compile-time known values_
+and _local compile-time known values_ is that the former also contains
+constructor parameters. The distinction is important when it comes to
+defining the meaning of features like types. For example, in the type
+`bit<e>`, the expression `e` must be a local compile-time known
+value. Suppose instead that `e` were a constructor parameter---i.e.,
+merely a compile-time known value. In this situation, it would be
+impossible to resolve `bit<e>` to a concrete type using purely local
+information---we would have to wait until the constructor was
+instantiated and the value of `e` known.
```

Comment:
Should this also include reasoning behind the generics being excluded? (I don't see a place where this would apply thought).

Reply:
This paragraph explains why we need to exclude constructor parameters from local compile-time known values. I was wondering if it should also mention the reasoning behind excluding generic arguments (i.e. " Expressions of the form `e.minSizeInBits()`, `e.minSizeInBytes()`, `e.maxSizeInBits()` and `e.maxSizeInBytes()` **where the the type of `e` is generic**"). It is necessary, I was just wondering if it would be nice. But it is a rather similar case so maybe it is redundant.